packages/backend/src/tokens/request.ts

```diff
@@ -688,15 +799,129 @@ ${error.getFullMessage()}`,
       );
     }
 
-    return signedOut(authenticateContext, err.reason, err.getFullMessage());
+    return signedOut({
+      tokenType: 'session_token',
+      authenticateContext,
+      reason: err.reason,
+      message: err.getFullMessage(),
+    });
+  }
+
+  function handleMachineError(tokenType: NonSessionTokenType, err: unknown): UnauthenticatedState<NonSessionTokenType> {
+    if (!(err instanceof MachineTokenVerificationError)) {
+      return signedOut({
+        tokenType,
+        authenticateContext,
+        reason: AuthErrorReason.UnexpectedError,
+      });
+    }
+
+    return signedOut({
+      tokenType,
+      authenticateContext,
+      reason: err.code,
+      message: err.getFullMessage(),
+    });
+  }
+
+  async function authenticateMachineRequestWithTokenInHeader() {
+    const { sessionTokenInHeader } = authenticateContext;
+    if (!sessionTokenInHeader) {
+      return handleError(new Error('No token in header'), 'header');
+    }
+
+    // Handle case where tokenType is any and the token is not a machine token
+    if (!isMachineToken(sessionTokenInHeader)) {
+      return signedOut({
+        tokenType: acceptsToken as NonSessionTokenType,
+        authenticateContext,
+        reason: AuthErrorReason.TokenTypeMismatch,
+        message: '',
+      });
+    }
+
+    const parsedTokenType = getMachineTokenType(sessionTokenInHeader);
+    const mismatchState = maybeHandleTokenTypeMismatch(parsedTokenType, acceptsToken, authenticateContext);
+    if (mismatchState) {
+      return mismatchState;
+    }
+
+    const { data, tokenType, errors } = await verifyMachineAuthToken(sessionTokenInHeader, authenticateContext);
+    if (errors) {
+      return handleMachineError(tokenType, errors[0]);
+    }
+    return signedIn({
+      tokenType,
+      authenticateContext,
+      machineData: data,
+      token: sessionTokenInHeader,
+    });
+  }
+
+  async function authenticateAnyRequestWithTokenInHeader() {
+    const { sessionTokenInHeader } = authenticateContext;
+    if (!sessionTokenInHeader) {
+      return handleError(new Error('No token in header'), 'header');
+    }
```

Comment:
Something seems weird about this logic...

Reply:
Yeah either that or we remove and do non-null assertions